pkg/api/kubegraph/analysis/hpa.go

```diff
@@ -0,0 +1,107 @@
+package analysis
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gonum/graph"
+
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	osgraph "github.com/openshift/origin/pkg/api/graph"
+	"github.com/openshift/origin/pkg/api/kubegraph"
+	kubenodes "github.com/openshift/origin/pkg/api/kubegraph/nodes"
+)
+
+const (
+	HPAMissingCPUTargetError      = "HPAMissingCPUTarget"
+	HPAOverlappingScaleRefWarning = "HPAOverlappingScaleRef"
+)
+
+// FindHPASpecsMissingCPUTargets scans the graph in search of HorizontalPodAutoscalers that are missing a CPU utilization target.
+// As of right now, the only metric that HPAs can use to scale pods is the CPU utilization, so if a HPA is missing this target it
+// is effectively useless.
+func FindHPASpecsMissingCPUTargets(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	for _, uncastNode := range g.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		hpa, ok := node.Object().(*extensions.HorizontalPodAutoscaler)
+		if !ok {
+			continue
+		}
+
+		if hpa.Spec.CPUUtilization == nil {
+			markers = append(markers, osgraph.Marker{
+				Node:     node,
+				Severity: osgraph.ErrorSeverity,
+				Key:      HPAMissingCPUTargetError,
+				Message:  fmt.Sprintf("%s is missing a CPU utilization target", node.Name),
+			})
+		}
+	}
+
+	return markers
+}
+
+// FindOverlappingHPAs scans the graph in search of HorizontalPodAutoscalers that are attempting to scale the same set of pods.
+// This can occur in two ways:
+//   - 1. label selectors for two ReplicationControllers/DeploymentConfigs/etc overlap
+//   - 2. multiple HorizontalPodAutoscalers are attempting to scale the same ReplicationController/DeploymentConfig/etc
+// Case 1 is handled by deconflicting the area of influence of ReplicationControllers/DeploymentConfigs/etc, and therefore we
+// can assume that it will be handled before this step. Therefore, we are only concerned with finding HPAs that are trying to
+// scale the same resources, exactly.
+func FindOverlappingHPAs(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+	references := map[extensions.SubresourceReference][]*kubenodes.HorizontalPodAutoscalerNode{}
+
+	for _, uncastNode := range g.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		hpa, ok := node.Object().(*extensions.HorizontalPodAutoscaler)
+		if !ok {
+			continue
+		}
+
+		references[hpa.Spec.ScaleRef] = append(references[hpa.Spec.ScaleRef], node)
+	}
+
+	for _, hpas := range references {
```

Comment:
This test isn't sufficient.  It only catches direct dependencies, not the fully set of overlaps.  If hpa1 references dc1 which references rc1 and hpa tries to control rc1, its still an overlap.